egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py

```diff
@@ -0,0 +1,1150 @@
+        f.close()
+
+        f = open(self.config['vars_path']+"/bias_one", 'w')
+        f.write(" [ ")
+        for i in range(cell_dim):
+            f.write("1 ")
+        f.write("]\n")
+        f.close()
+
+        # formulation for OPGRU like:
+        # z_t = \sigmoid ( x_t * U^z + s_{t-1} * W^z ) // update gate
+        # o_t = \sigmoid ( x_t * U^o + s_{t-1} * W^o ) // output gate
+        # \tilde{h}_t = \tanh ( x_t * U^h + y_{t-1} \dot W^h ) // W^h is learnable vector
+        # h_t = ( 1 - z_t ) \dot h_t + z_t \dot y_{t-1}
+        # y_t = ( y_t \dot o_t) * W^y
+        # s_t = y_t(0:rec_proj_dim-1)
+
+        
+        configs = []
+        configs.append("# Update gate control : W_z* matrics")
+        configs.append("component name={0}.W_z.xs_z type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + rec_proj_dim, cell_dim, affine_str))
+        
+        configs.append("# Output gate control : W_r* matrics")
+        configs.append("component name={0}.W_z.xs_o type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + rec_proj_dim, cell_dim, affine_str))
+
+        configs.append("# h related matrix : W_h* matrics")
+        configs.append("component name={0}.W_h.UW type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim , cell_dim , affine_str))
+        configs.append("component name={0}.W_h.UW_elementwise type=NaturalGradientPerElementScaleComponent dim={1} {2}".format(name, cell_dim , pes_str))
+        
+        configs.append("# Defining the non-linearities")
+        configs.append("component name={0}.z type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.o type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.h type=TanhComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+
+        configs.append("# Defining the components for other cell computations")
+        configs.append("component name={0}.o1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y2 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y type=NoOpComponent dim={1}".format(name, cell_dim))
+
+        configs.append("# Defining fixed scale/bias component for (1 - z_t)")
+        configs.append("component name={0}.fixed_scale_minus_one type=FixedScaleComponent scales={1}".format(name, self.config['vars_path']+"/minus_one"))
+        configs.append("component name={0}.fixed_bias_one type=FixedBiasComponent bias={1}".format(name, self.config['vars_path']+"/bias_one"))
```

Comment:
It should not be necessary to use fixed scale and bias now that I have merged those upgrades to Descriptors.   Can you upgrade these scripts?

Also, do you have any plan to implement a fast version of the GRU comopnent?

Reply:
Will it be OK to change the config into Descriptors but use the exps ran by scale-bias ?
It's time consuming to rerun all the scripts.
I'm not sure when I am OK to update a fast version OPGRU (it may need some tuning), I'm a little busy this month. I will ask Huang Lu whether he is OK.

Reply:
yes that's OK, it's totally equivalent-- just run something small to make sure it doesn't crash and the objectives aren't wildly different than before for the 1st iteration.